ldap/servers/plugins/acctpolicy/acct_plugin.c

```diff
@@ -208,75 +209,96 @@ acct_update_login_history(const char *dn, char *timestr)
             }
-            /* remap array*/
-            for (i = 0; i < (cfg->login_history_size - 1); i++) {
-                login_hist[i] = login_hist[(diff + 1) + i];
+            /* remap array if it exists */
+            for (i = 0; i < cfg->login_history_size; i++) {
+                login_hist[i] = login_hist[diff + (i + 1)];
+            }
+            if (cfg->login_history_size != 0) {
+                /* realloc and append latest value */
+                login_hist = (char **)slapi_ch_realloc((char *)login_hist, sizeof(char *) * (cfg->login_history_size + 2));
+                login_hist[i - 1] = slapi_ch_smprintf("%s", timestr);
+                login_hist[i] = NULL;
+                mod_required = 1;
+            } else {
+                /* size is 0 and array has been trimmed, we need a mod */
+                mod_required = 1;
             }
-            /* expand array and add current time string at the end */
-            login_hist = (char **)slapi_ch_realloc((char *)login_hist, sizeof(char *) * (cfg->login_history_size + 1));
-            login_hist[i] = slapi_ch_smprintf("%s", timestr);
-            login_hist[i + 1] = NULL;
         } else {
-            /* expand array and add current time string at the end */
-            login_hist = (char **)slapi_ch_realloc((char *)login_hist, sizeof(char *) * (num_entries + 2));
-            login_hist[num_entries] = slapi_ch_smprintf("%s", timestr);
-            login_hist[num_entries + 1] = NULL;
+            if (cfg->login_history_size != 0) {
+                /* realloc and append latest value */
+                login_hist = (char **)slapi_ch_realloc((char *)login_hist, sizeof(char *) * (num_entries + 2));
+                login_hist[num_entries] = slapi_ch_smprintf("%s", timestr);
+                login_hist[num_entries + 1] = NULL;
+                mod_required = 1;
+            }
         }
+        /* first time round */
+    } else if (cfg->login_history_size > 0) {
+        login_hist = (char **)slapi_ch_calloc(2, sizeof(char *));
+        /* alloc new array and append latest value */
+        login_hist = (char **)slapi_ch_realloc((char *)login_hist, sizeof(char *) * (num_entries + 2));
+        login_hist[num_entries] = slapi_ch_smprintf("%s", timestr);
+        login_hist[num_entries + 1] = NULL;
+        mod_required = 1;
     }
 
-    /* modify the attribute */
-    attribute.mod_type = cfg->login_history_attr;
-    attribute.mod_op = LDAP_MOD_REPLACE;
-    attribute.mod_values = login_hist;
-
-    list_of_mods[0] = &attribute;
-    list_of_mods[1] = NULL;
-
-    mod_pb = slapi_pblock_new();
-    slapi_modify_internal_set_pb(mod_pb, dn, list_of_mods, NULL, NULL, plugin_id, 0);
-    slapi_modify_internal_pb(mod_pb);
-    slapi_pblock_get(mod_pb, SLAPI_PLUGIN_INTOP_RESULT, &rc);
-    if (rc != LDAP_SUCCESS) {
-        slapi_log_err(SLAPI_LOG_ERR, "acct_update_login_history", "Modify error %d on entry '%s'\n", rc, dn);
+    if (mod_required) {
```

Comment:
I have a doubt, if login_history_size==0 what will be the result ?
An attribute 'lastloginhistory' present with no value, or no 'lastloginhistory' attribute ?
I think it would be better to have no 'lastloginhistory' attribute.

Reply:
No problems here: 
   in a modify replace a NULL or an empty modifier list means that the attribute is removed. 
BTW in ldap there is no such thing as an attribute without value. 
 ( Attribute always have a value (that value may be empty but in this case you have to specify explicitly that a value is present and the value size is 0)

Reply:
> I have a doubt, if login_history_size==0 what will be the result ?

If login_history_size == 0 then the feature is disabled and we do nothing unless the entries array has been trimmed.